netebpfext/net_ebpf_ext_sock_addr.c

```diff
@@ -423,7 +839,479 @@ net_ebpf_extension_sock_addr_authorize_connection_classify(
+    uint32_t verdict,
+    _In_ const bpf_sock_addr_t* original_context,
+    _In_ const bpf_sock_addr_t* redirected_context,
+    _In_ const FWPS_FILTER* filter,
+    uint64_t classify_handle,
+    HANDLE redirect_handle,
+    _Out_ bool* redirected,
+    _Inout_ FWPS_CLASSIFY_OUT* classify_output)
+{
+    NTSTATUS status = STATUS_SUCCESS;
+    FWPS_CONNECT_REQUEST* connect_request = NULL;
+    bool commit_layer_data = false;
+
+    *redirected = false;
+
+    if (verdict == BPF_SOCK_ADDR_VERDICT_PROCEED) {
+        // Check if destination IP and/or port have been modified.
+        bool address_changed = !_net_ebpf_ext_compare_destination_address(redirected_context, original_context);
+        if (redirected_context->user_port != original_context->user_port || address_changed) {
+            *redirected = true;
+
+            status = FwpsAcquireWritableLayerDataPointer(
+                classify_handle, filter->filterId, 0, (PVOID*)&connect_request, classify_output);
+            if (!NT_SUCCESS(status)) {
+                NET_EBPF_EXT_LOG_NTSTATUS_API_FAILURE_UINT64_UINT64(
+                    NET_EBPF_EXT_TRACELOG_KEYWORD_SOCK_ADDR,
+                    "FwpsAcquireWritableLayerDataPointer",
+                    status,
+                    filter->filterId,
+                    (uint64_t)redirected_context->compartment_id);
+
+                goto Exit;
+            }
+            commit_layer_data = true;
+
+            InterlockedIncrement(&_net_ebpf_ext_statistics.redirect_connection_count);
+
+            if (redirected_context->user_port != original_context->user_port) {
+                INETADDR_SET_PORT((PSOCKADDR)&connect_request->remoteAddressAndPort, redirected_context->user_port);
+            }
+            if (address_changed) {
+                uint8_t* address;
+                if (redirected_context->family == AF_INET) {
+                    address = (uint8_t*)&redirected_context->user_ip4;
+                } else {
+                    address = (uint8_t*)&(redirected_context->user_ip6[0]);
+                }
+                INETADDR_SET_ADDRESS((PSOCKADDR)&connect_request->remoteAddressAndPort, address);
+            }
+
+            // Target process id and local redirect handle needs to be set in two cases:
+            // 1. The redirected address is loopback.
+            // 2. The redirected address is a local non-loopback address.
+            // To simplify the design, always set these values.
+            connect_request->localRedirectTargetPID = TARGET_PROCESS_ID;
+            connect_request->localRedirectHandle = redirect_handle;
+        } else {
+            InterlockedIncrement(&_net_ebpf_ext_statistics.permit_connection_count);
+        }
+    } else {
+        InterlockedIncrement(&_net_ebpf_ext_statistics.block_connection_count);
+    }
+
 Exit:
+    if (commit_layer_data) {
+        FwpsApplyModifiedLayerData(classify_handle, connect_request, 0);
+    }
+
+    return status;
```

Comment:
use `NET_EBPF_EXT_RETURN_RESULT`. Please make this change elsewhere as needed.

Reply:
fixed at all places.